# Virome_2021  
- There are codes used in **"Hidden viral infections in public sequencing data and warning for future viral zoonotic diseases"**.  
- DOI: 10.1101/2021.05.17.444395  


**  **
## Table of Contents  
1. **Virome_sequence_assembly**  
The details were described in the section: *"Sequence assembly using publicly available RNA-seq data"* in our paper.  


2. **Virome_extract_virus_contigs**  
The details were described in the section: *"Identification of contigs originating from RNA viruses"* in our paper.  


3. **Virome_totalize_virus_infections**  
The details were described in the section: *"Totalization of RNA viral infections in public RNA-seq data"* in our paper.  


4. **Virome_characterize_virus_genome**  
The details were described in the section: *"Characterization for viral genomic architectures"* in our paper.  


5. **Virome_phylogenetic_analyses**  
The details were described in the section: *"Phylogenetic analyses"* in our paper.  


6. **Virome_mapping_analyses**  
The details were described in the section: *"Mapping analyses using viral genomes identified in this study"* in out paper.  

**  **  
## Bioinformatics tools and their versions  
Bioinformatics tools and their versions are listed in **"Bioinformatics_tools_virome2021.tsv**.  


**  **
## Datasets in the Mendeley Data  
The relevant row data are available at the [Mendeley Data](http://dx.doi.org/10.17632/stscmh9mr3.1).  

- RNA_viral_contigs.fasta  
These sequences were obtained according to the procedures described in *"Sequence assembly using publicly available RNA-seq data"* in our paper.  

- Riboviria_viral_protein.fasta  
These sequences were used for the BLASTX screening to extract RNA viral sequences. The detailed procedures were described in *"Identification of contigs originating from RNA viruses"* in our paper.  

- Alignment_coverage.txt  
The viral infections identified in this study were listed in the table.  The detailed procedures were described in *"Totalization of RNA viral infections in public RNA-seq data"* in our paper.  

